### DISCUSSION TITLE
Using scopes on `Security()` changes dependency inheritance

### DISCUSSION THREAD

**Discussion by TheDJVG** [fastapi/fastapi#8421] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [x] I commit to help with one of those options 👆

### Example Code

```python
async def get_authenticated_user(security_scopes: SecurityScopes,
                                 session: Session = Depends(get_session),
                                 token: str = Depends(oauth2_scheme),
                                 ):

    print(f"get_authenticated_user: {id(session)}")
    return

@router.post('/activate', status_code=204)
async def user_activate(session: Session = Depends(get_session),
                        user: UserAccountInDatabase = Security(
                            get_authenticated_user,
                            scopes=["activate"])):
    print(f"user_activate: {id(session)}")
    return
```

### Description

As you can see I have `scopes=["activate"]` defined in this `Security()` example.

The result will print:
```
get_authenticated_user: 140600933628496
user_activate: 140600933312848
```

As you can see the IDs of both sessions are different, this is unexpected to me as I was under the impression that arguments propagate through dependencies.

The weird thing is that when I remove `scopes=["activate"]` from `Security()` I do get the expected result:
```
get_authenticated_user: 139952151843088
user_activate: 139952151843088
```

I figured this out when I tried to commit a change in the `UserAccountInDatabase` object and sqlalchemy complained that the object does not belong to the same `Session`.

I'm wondering if this has anything to do with the instantiation of a SecurityScopes at some point by Security(). I tried manually adding the security scopes like this:
```python
activation_security_scope = SecurityScopes(["activate"])
@router.post('/activate', status_code=204)
async def user_activate(session: Session = Depends(get_session),
                        security_scopes: SecurityScopes = activation_security_scope,
                        user: UserAccountInDatabase = Depends(
                            get_authenticated_user)):
    print(f"user_activate: {id(session)}")
    print(security_scopes.scope_str)
```

But this actually will just show that `security_scopes` is empty, it almost seems like it's being overwritten somewhere (maybe by the `Depends(get_authenticated_user)`. 

As I'm new to FastAPI there's a big chance I'm misunderstanding something.

### Operating System

Linux

### FastAPI Version

0.87.0

### Python Version

3.11.0

**Comment by sehraramiz** [comment, marked as answer]
[Dependant](https://github.com/tiangolo/fastapi/blob/63a5ffcf577abc6015d021f65857b3826d8db74e/fastapi/dependencies/models.py#L15) object has a cache_key attribute which the dependency resolver of fastapi uses it to cache dependencies and use the already resolved dependencies (get_session in this case).
according to [this line](https://github.com/tiangolo/fastapi/blob/63a5ffcf577abc6015d021f65857b3826d8db74e/fastapi/dependencies/models.py#L58) of Dependant class the cache_key is generated using the **dependency function reference** and **security scopes**.
when you use get_authenticated_user as a security dependency every param in get_authenticated_user gets the security scopes of the parent api (user_activate).
so when resloving dependencies from user_activate with scopes, the param ```session: Session = Depends(get_session)``` gets saved in dependency_cache with a key like ```(<function get_session at 0x7fb87827c9d0>, ('activate'))```
when resloving dependencies from user_activate with no scopes, the param ```session: Session = Depends(get_session)``` gets saved in dependency_cache with a key like ```(<function get_session at 0x7fb87827c9d0>, ())```

i think this behavior might be a security feature.

it might help you to read [this part of the doc](https://fastapi.tiangolo.com/tutorial/dependencies/sub-dependencies/?h=dependenc#using-the-same-dependency-multiple-times) on dependency cache
and [this part](https://fastapi.tiangolo.com/advanced/security/oauth2-scopes/?h=dependency#dependency-tree-and-scopes) about dependency tree and scopes.

**Comment by ShellCode33** [comment]
I faced the same issue and ended up using `request.state` to manually cache the session instead of relying on the built-in cache mechanism.

It's very similar to the example from op, but just in case here's the minimal reproducible example I came up with:


```python
from typing import Annotated

from fastapi import Depends, FastAPI, Security

app = FastAPI()


class DbSession:
    pass


def get_db() -> DbSession:
    return DbSession()


async def check_permissions(db_session: Annotated[DbSession, Depends(get_db)]) -> None:
    print("db_session id (from check_permissions):", id(db_session))


async def path_dep(db_session: Annotated[DbSession, Depends(get_db)]) -> str:
    print("db_session id (from path_dep):", id(db_session))
    return "whatever"


@app.get("/", dependencies=[Security(check_permissions, scopes=["test"])])
# @app.get("/", dependencies=[Security(check_permissions)])
async def root(
    db_session: Annotated[DbSession, Depends(get_db)],
    param: Annotated[str, Depends(path_dep)],
):
    print("db_session id (from root):", id(db_session))
    print("param:", param)
    return {"message": "Hello World"}
```

Try to reach the root route, and check out the object IDs, they are different:

```
db_session id (from check_permissions): 140207666429648
db_session id (from path_dep): 140207666416144
db_session id (from root): 140207666416144
```

Now try to remove scopes from the Security dependency, IDs are now the same. While I understand that scopes must be considered when caching Security dependencies, I feel like there's something wrong with the way sub-dependencies are handled. The `db_session` parameter is a `Depends` dependency, not a `Security` one. I don't understand why the fact that it comes from a `Security` dependency alters the caching behaviour of sub-dependencies.

What are your thoughts on this @tiangolo ?

**Comment by shughes-uk** [comment]
Discovered this today. It's quite a nuisance in practice.

Let's say you have a sqlalchemy session dep `get_db` so you can share a single transaction/session across the lifetime of a request.

Then you have another `Security` dep that pulls auth info and then grabs the `User` from the db , `current_user` or somesuch.

Now in your endpoint you define your deps like

```python
user=Security(current_user)
session=Depends(get_db)
```

Your `user` object will be detatched from the `session`, and you'll also have used two connections instead of a single one.

**Comment by qmorek** [comment]
Related: https://github.com/fastapi/fastapi/discussions/6024#discussioncomment-8541913

**Reply by YuriiMotov** [reply]
Wrong link or wrong place :)

**Reply by qmorek** [reply]
Wrong link, thanks for pointing it out 😄
